### DISCUSSION TITLE
Localization

### DISCUSSION THREAD

**Discussion by exectails** [NoCode-NoLife/melia#174] [Discussions]
While we'll probably never need to localize any official strings, as all dialogues and notices can be found in the client's localization dictionary, there will always be situations where we send custom strings to the client, be it for a custom NPC, a custom notice, or GM command results. Since we can't utilize the client's localization in this case we would need our own way to localize them, provided that we'd want that.

Does Melia need a localization system for its non-official strings, or will the few custom messages and NPCs we will have in master, that admins didn't code in their own language to begin with, be fine in English?

In Aura we're using a [gettext-like system](https://github.com/aura-project/aura/tree/master/system/localization) that I like, but unlike Melia Aura actually needs to localize every string it sends to the clients itself.

It's better to decide this sooner than later, as every custom string added to the source or custom NPCs means one more string we have to update later on if we decide to add localization.

**Comment by raiken-mf** [comment]
I don't think localization is needed, since the client is Korean or English, just some parts in for example German would look ... weird, also if something important needs to be translated some custom NPC or a second broadcast in the specific language would do the job.

**Comment by exectails** [comment]
@lRaikeNl I understand what you mean, but the problem is, we do have custom messages, and we do have custom NPCs, both of which will increase in number over time. Those will then be in English, while the user might set their client to Korean, or Portuguese, or Japanese, or god knows what localizations will follow.

We have a general problem here, with the client handling the entire localization officially, because we'll basically never be able to ensure that a user sees *only* their locally chosen language. Even if we have server-side localization, and even if our community was to provide translations for everything in multiple languages, the server wouldn't know which language the user chose locally, and would always send, for example, English.

There's also the matter of #103, and us potentially not using the client-sided localization for NPC dialogues anymore, which would make this problem worse.

**Comment by raiken-mf** [comment]
Since it isn't possible to ensure what language the user sets, is it possible to set the users language on the server side? (On the client)

**Comment by exectails** [comment]
Not that I know of.

**Comment by celophi** [comment]
The client sends the locale of the OS upon login.
Additionally, there is a setting to enable/disable the use of dictionary translation.

**Comment by exectails** [comment]
You can't really depend on the OS locale, can you? I have a German Windows for example, but I always play MMOs in English if given the option. Which packet is that information sent in though?

**Comment by celophi** [comment]
> We have a general problem here, with the client handling the entire localization officially, because we'll basically never be able to ensure that a user sees only their locally chosen language.

Simply allow the user to manage a language preference through a website or other interface.

**Comment by raiken-mf** [comment]
Or maybe through a NPC Script at the start location. But that should be no thing we should think about if the text messages are sent from server side. I'd prefer sending text messages from the server side. Even though the client knows the messages it would be a mess if a NPC Script would change behavior or text from official updates.

**Comment by exectails** [comment]
>Simply allow the user to manage a language preference

That would be an option. Although it's a little more complex because you then have to craft player specific packets based on their language settings. If you have an announcement that goes to multiple characters for example, you have to make sure they all get the message in their selected language.

>it would be a mess if a NPC Script would change behavior or text from official updates

Yea, the NPC convention issue we have is definitely related to this (#103). I imagine only ever working with localization keys would be really annoying for scripters or people who simply want to understand what's happening in a script. And when a user is looking for a specific NPC they might try to search by dialog, but that wouldn't be possible with keys.

If ToS didn't have client side NPC dialog this wouldn't even be up for discussion... we'd simply add a localization system, write out the dialog, and be done with it. And if the player chooses a server that's in a language they don't understand that's on them. (Unless we did selectable languages, but that would not be pretty.)

**Comment by exectails** [comment]
By now we decided not to rely on client-sided dialog lookup, which means that we need localization support for language selection of any kind. Case closed.
